core/src/main/java/org/mule/runtime/core/api/util/MessagingExceptionResolver.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+
+package org.mule.runtime.core.api.util;
+
+import static java.util.stream.Collectors.toList;
+import static org.mule.runtime.api.exception.ExceptionHelper.getExceptionsAsList;
+import static org.mule.runtime.core.api.context.notification.EnrichedNotificationInfo.createInfo;
+import static org.mule.runtime.core.api.util.ExceptionUtils.createErrorEvent;
+import static org.mule.runtime.core.api.util.ExceptionUtils.getComponentIdentifier;
+import static org.mule.runtime.core.api.util.ExceptionUtils.getErrorMappings;
+import static org.mule.runtime.core.api.util.ExceptionUtils.getMessagingExceptionCause;
+import static org.mule.runtime.core.api.util.ExceptionUtils.isUnknownMuleError;
+
+import org.mule.runtime.api.component.ComponentIdentifier;
+import org.mule.runtime.api.event.Event;
+import org.mule.runtime.api.message.Error;
+import org.mule.runtime.api.message.ErrorType;
+import org.mule.runtime.core.api.InternalEvent;
+import org.mule.runtime.core.api.MuleContext;
+import org.mule.runtime.core.api.context.notification.EnrichedNotificationInfo;
+import org.mule.runtime.core.api.exception.ErrorTypeLocator;
+import org.mule.runtime.core.api.exception.MessagingException;
+import org.mule.runtime.core.api.message.ErrorBuilder;
+import org.mule.runtime.core.api.processor.Processor;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * Tries to resolve a new {@link MessagingException}s with the real cause of the problem from an incoming
+ * {@link MessagingException} that carries a list of causes inside it so it can be thrown and no information is lost.
+ *
+ * @since 4.0
+ */
+public class MessagingExceptionResolver {
+
+  /**
+   * Resolves a new {@link MessagingException} with the real cause of the problem based on the content of an
+   * Incoming {@link MessagingException} with a chain of causes inside it and the current event that the exception is carrying.
+   * <p>
+   * This method will pick the FIRST cause exception that has a mule or extension KNOWN error as the real cause, if there is not
+   * an exception in the causes that match with an Known error type then this method will try to find the error that the current
+   * {@link Event} is carrying.
+   * <p>
+   * When there are multiple exceptions that contains the same root error type, then this method will wrap the one that has
+   * highest position in the causes list
+   *
+   * @return a {@link MessagingException} with the proper {@link Error} associated to it's {@link InternalEvent}
+   */
+  public static MessagingException resolve(Processor processor, MessagingException me, MuleContext context) {
```

Comment:
having this as a static method might be convenient, but prevents this from being mock and build tests in which a specific exception is resolved. 

Consider turning this into an instance method. If you have a good reason not to, then this class needs to be final and have a private constructor